### PR TITLE
Double-scanning support for VESA modes, accurate mode change logging, `vga_draw.cpp` rework & many other improvements

### PR DESCRIPTION
**Pull request by johnnovak**
# Introduction

If it wasn't _absolutely clear_ to everyone here that I'm a man who does not believe in half-measures, this PR should really drive that home! 😎 🚬 

This started as something quite innocent: I only wanted to expand upon kcgen's excellent `vgaonly` work to add proper double-scanning to all VESA modes (see this ticket https://github.com/dosbox-staging/dosbox-staging/issues/2604)  because it was annoying me that late DOS era demos that use high-colour modes appeared single-scanned... like an EGA monitor 😬 Plus I wanted to fix the logging on video mode changes so it logs the dimensions of the actual video mode, not the render buffer.

Sounds easy right? It turns out it was a case of having to take 10 steps back and take a quite different route just to be able to take _one step_ forward. I discussed this with kcgen, his work was mostly based on heuristics and patching things up here and there, but that approach wasn't really maintainable and scalable; continuing down that track would've resulted in a huge unmaintainable mess. Still, without him bravely forging the path ahead, I would never have dared to touch the scary beast that is `vga_draw.cpp`! 

So one thing lead to another, and suddenly I've sunk easily 150+ hours into this over the last month or so... Running literally thousands of test cases (see test cases below), fun stuff like that 😎 I'm still in recovery mode, refactoring this _was_ exhausting...

To make sense half of all this, you need to be quite familiar with CGA/EGA/VGA... especially with VGA, the custom "Mode X" tweak mode stuff, and how double-scanning of VGA adapters affect image quality on real CRT monitors. Otherwise, you'll just need to trust me 😎 🤘🏻 

I've tested things quite extensively over and over again, so I'm quite confident in these changes, but of course it's not impossible that some edge cases are broken. However, with the new, better structure in place, it will be _much_ easier to fix problems with _one_ specific video mode on _one_ adapter without breaking random stuff on other adapters. That's not the case in current main...

# Overview

Ok, so main changes in a nutshell (well, it's a rather large nut... 😅 🥜 )

**REVIEW BY COMMIT IF YOU VALUE YOUR SANITY!!! I've put considerable effort in coming up with a relatively logical and clean history (I must have spent several full days just reorganising stuff with rebase...)**

## Double-scanning 

- The `force_vga_single_scan` setting has been moved to the `[render]` `section.

- `force_vga_single_scan` can now be toggled at runtime and it gets activated on the next video mode change (which I find pretty cool).

- *All* VESA modes supported by DOSBox are now double-scanned according to the mode table in `int10_modes.cpp` (tested with `vesa_modes = all` and `vmemsize = 8`). Naturally, the same goes for all non-VESA SVGA modes too on all supported SVGA adapters.

- CGA, PCjr & Tandy text modes are now properly single-scanned.

- The determination of whether a given VGA mode should be double-scanned or not should be almost perfectly correct now (I just don't dare tempt the devil by saying it's 100% 😈 🔥 ). This is because double-scanning is now controlled by the state of VGA registers for the most part instead of various image dimension and BIOS mode based heuristics. This has been tested with demos and various emulators (RealSpec, CCS64) that set up wacky custom modes to for example achieve a single-scanned look with around 300-lines of resolution at 85Hz+ refresh rates... Yep, all that works, in addition to all the games and demoscene productions that use strange custom modes on VGA!

## Pixel-doubling

- Sub-640 pixel wide video modes are always pixel-doubled now by default. This is needed to achieve more accurate CRT emulation with virtually all current CRT shaders in existence. For example, on a real monitor there should be *zero* difference between a 320x200 image displayed in a 320x200 video mode, or the same image width-doubled and displayed in a 640x200 mode. Without pixel-doubling, we would have a very noticeable difference with current CRT shaders that don't take such differences into account, resulting in horizontally blurrier image for the 320x200 video mode case (there are some image comparisons on this below).

- A new `force_no_pixel_doubling` render setting is also available to disable this behaviour. This might be useful for optimising performance on low-powered devices, or to emulate a really blurry consumer CRT monitor or TV set.

- See these threads for my discussions with prominent shader authors / people
  with an intense interest in accurate CRT emulation on the matter:

  - Discussion on Vogons with lots of interesting technical detail about analog monitors:
  https://www.vogons.org/viewtopic.php?f=63&t=94909
  - My question to Guest (author of the "CRT Advaned Guest" shader, one of the best CRT shaders around): 
  https://forums.libretro.com/t/new-crt-shader-from-guest-crt-guest-advanced-updates/25444/3591?u=rincewind
  - Guest's very informative reply:
  https://forums.libretro.com/t/new-crt-shader-from-guest-crt-guest-advanced-updates/25444/3592?u=rincewind

- VileR was kind enough to confirm the above by taking some actual photos of his CGA monitor in 320x200 and 640x200 modes: https://www.vogons.org/viewtopic.php?p=1174367#p1174367

- The TL;DR is that pixel-doubling of roughly sub-640 pixel wide content is necessary to achieve an authentic CRT look with currently available CRT shader technology, without having to maintain a set of different shaders and switch between them manually depending on whether you're playing a game that uses ~320 or ~640 pixel wide modes. Plus I find the multi-shader approach a non-solution when using DOSBox as a computer emulator because a shader optimised for ~320-pixel-wide content would look off in ~640-pixel-wide modes, and vice versa.

## Shaders

- Introduced two new shader pragmas, `force_single_scan` and `force_no_pixel_doubling`, so shaders can explicitly request single-scanning and no pixel-doubling if this would make no visual difference (we cannot determine this upstream, only the shader author can make this decision). The default "sharp" shader is such an example; this allows for finer scaling granularity with integer scaling enabled, and there are 4x less pixels to process by the shader which can make a big difference on low-powered devices.

- The default `interpolation/sharp` shader has been updated to force single-scanning and no pixel-doubling. This ensures there's _zero_ negative performance impact on people who are content with just using the defaults.

- These pragmas also for cool possibilities, e.g. to make a "Commodore monitor" shader that forces single-scanning and no pixel-doubling.

## Logging of current video mode

- The removal of `vgaonly` broke the logging of the video mode dimensions because only the rendered image dimensions were available downstream, so we got things like `MCGA 640x400 (mode 13h)` in the logs which is super-confusing and incorrect. That is now fixed; e.g., for mode 13h we get `VGA 320x200 (mode 13h)` as we should.

- The logging of video modes is now a _lot_ more accurate in general; now we're consistently logging the video standard the mode _first_ appeared in (e.g., CGA modes should be always CGA, regardless of the emulated display adapter in use).

- If there's ambiguity regarding the video mode (e.g., the actual source image dimensions don't match the nominal dimensions of the BIOS video mode), we omit the `(mode XYh)` part from the logs to avoid confusing the user. The most noticeable side effect of this is that we no longer log weird stuff like `MCGA 360x240 (mode 13h)` which is clearly not mode 13h but a tweaked VGA mode---it's just mode 13h was used as a _starting point_ to tweak the mode further. Stuff like that gets simply logged as `VGA 360x240` now.

- Another subtle but important detail: now the pixel aspect ratio of the _video mode_ gets logged, not whatever pixel aspect ratio we derive from the _rendered image_'s dimensions for rendering purposes.

- Composite CGA/Tandy/PCjr modes in particular are being logged more accurately, and we get video mode change log message when toggling composite modes.

## Image capture

_If the below points don't make much sense, it's time to refresh your memory about the new enhanced & revamped image capture feature: https://github.com/dosbox-staging/dosbox-staging/pull/2503_

- Because now the dimensions and PAR of the video mode get propagated downstream, the image capturer can write the correct metadata tags into the images (this was just as incorrect as the logging of the current video mode's dimensions).

- The `raw` image capture mode now _really_ means raw; e.g., a 640x200 image isn't height-doubled like before but simply written as a 640x200 pixel-raster, end of story. This is the logical conclusion of the new revamped image capture functionality; the `upscaled` aspect-ratio preserving, "sharp" upscaled output is the default, this is what most people should use. The `raw` mode that corresponded to the "legacy" DOSBox screenshot capture functionality break compatibility with it on purse; this mode is really just there for special needs when you must capture the original, unaltered pixel-raster data.

- The image capturer now always uses the dimensions of the _video mode_ as the basis of performing the upscaling (as opposed to the dimensions of the _rendered image_, as done previously). In some cases (when dealing with "baked in" double-scanning) this can result in hitting the target ~1200px of upscaled vertical resolution better (because we can effectively use 2x finer vertical integer scaling steps for "baked-in" double-scanning).

- Naturally, and following from the previous point, all combinations of pixel-doubling, "fake" double-scanning, real "baked-in" double-scanning are handled 100% correctly as you would expect during image capturing. Basically, the results for the same video mode should always be completely identical, regardless of these rendering side variations.

### CODE REVIEW

**Comment by johnnovak** [comment]
# Testing details

## Deluxe Paint IIe

- All supported video modes on all cards (starting from Hercules) with all double-scan, pixel-doubling & aspect ratio correction flag permutations. This includes non-VESA SVGA modes too on the emulated Paradise and Tseng cards.

## QPV

- All S3 VGA and VESA modes with all double-scan, pixel-doubling & aspect ratio correction flag permutations. This includes all 15/16-bit high colour and 24-bit and 32-bit true colour modes.

## Text modes

- All text modes on all emulated adapters, including 132-column VESA text modes.

## Games

### Hercules

- California Games
- Total Eclipse

### CGA

- California Games
- Life and Death
- Romance of the Three Kingdoms
- Total Eclipse

### CGA/Tandy/PCjr composite

See https://github.com/dosbox-staging/dosbox-staging/wiki/Composite-mode-test-cases

- Adventures in Serenia
- Ancient Art of War, The
- Bruce Lee
- California Games
- Maniac Mansion
- Maniac Mansion Enhanced
- Zak McKracken Enhanced

### PCjr

- Adventures in Serenia
- King's Quest 1

### Tandy

- Bruce Lee
- California Games
- Maniac Mansion Enhanced
- Sargon 5 - World Class Chess
- Star Trek - 25th Anniversary
- Tandy Corp Christmas Demo
- Time Riders in American History
- Total Eclipse
- Zak McKracken Enhanced

### EGA/VGA

Games that use either custom resolutions or weird VGA tricks (e.g. to change the palette mid-screen, like Lemmings in VGA mode). See the big spreadsheet below for the list of tested custom resolutions.

- Alien Carnage
- Axia
- Bananoid
- Blood and Magic
- Copper
- CopperFake
- Covox Lemmings
- Cyril Cyberpunk
- Die Hoehlenwelt Saga
- Donut Demo
- Earthworm Jim
- Flashback
- Get Saddam
- Good, Bad, Ugly
- Holiday Lemmings
- Incredible Machine 1, The
- Incredible Machine 2, The
- Jazz Jackrabbit Holiday Hare
- Jump 'n Bump
- Jurassic Park
- Lost Vikings, The
- Magic Pockets
- Mario and Luigi
- Master Blaster
- Moraffs Monster Memory
- Moraffs Super Entrap
- Pinball Fantasies
- Pinball Illusions
- Prehistorik 2
- Ravenloft - Stone Prophet
- Ravenloft - Strahd's Possession
- Scorched Earth
- Summoning, The
- System Shock
- Threat
- Three Kingdoms
- Xiphos
- Xmas Lemmings 1991
- Xmas Lemmings 1992
- Yo! Joe!

## Test sheet

Including details about the upscaled image captures.

- PDF version: [DOSBox Staging image capture upscaling.pdf](https://github.com/dosbox-staging/dosbox-staging/files/12060433/DOSBox.Staging.image.capture.upscaling.pdf)
- ODS spreadsheet:  [DOSBox Staging image capture upscaling.ods.zip](https://github.com/dosbox-staging/dosbox-staging/files/12060462/DOSBox.Staging.image.capture.upscaling.ods.zip)

image

**Comment by johnnovak** [comment]
## Effects of pixel-doubling

Open both in two browser tabs and switch between them at 100% magnification to appreciate the differences.

### With pixel-doubling

EGA 320x200 rendered with pixel-doubling, so 640x200 is sent to the shader. The resulting image looks like a typical sharp EGA monitor.

image


### With no pixel-doubling

EGA 320x200 rendered with _no_ pixel-doubling, so 3200x200 is sent to the shader. The result looks like a cheap 15kHz monitor or a consumer TV set (too much horizontal blur; EGA monitors were meant for business applications and were capable of displaying sharp 640x200 text and graphics).

image

**Comment by johnnovak** [comment]
# Testing details

## Deluxe Paint IIe

- All supported video modes on all cards (starting from Hercules) with all double-scan, pixel-doubling & aspect ratio correction flag permutations. This includes non-VESA SVGA modes too on the emulated Paradise and Tseng cards.

## QPV

- All S3 VGA and VESA modes with all double-scan, pixel-doubling & aspect ratio correction flag permutations. This includes all 15/16-bit high colour and 24-bit and 32-bit true colour modes.

## Text modes

- All text modes on all emulated adapters, including 132-column VESA text modes.

## Games

### Hercules

- California Games
- Total Eclipse

### CGA

- California Games
- Life and Death
- Romance of the Three Kingdoms
- Total Eclipse

### CGA/Tandy/PCjr composite

See https://github.com/dosbox-staging/dosbox-staging/wiki/Composite-mode-test-cases

- Adventures in Serenia
- Ancient Art of War, The
- Bruce Lee
- California Games
- Maniac Mansion
- Maniac Mansion Enhanced
- Zak McKracken Enhanced

### PCjr

- Adventures in Serenia
- King's Quest 1

### Tandy

- Bruce Lee
- California Games
- Maniac Mansion Enhanced
- Sargon 5 - World Class Chess
- Star Trek - 25th Anniversary
- Tandy Corp Christmas Demo
- Time Riders in American History
- Total Eclipse
- Zak McKracken Enhanced

### EGA/VGA

Games that use either custom resolutions or weird VGA tricks (e.g. to change the palette mid-screen, like Lemmings in VGA mode). See the big spreadsheet below for the list of tested custom resolutions.

- Alien Carnage
- Axia
- Bananoid
- Blood and Magic
- Copper
- CopperFake
- Covox Lemmings
- Cyril Cyberpunk
- Die Hoehlenwelt Saga
- Donut Demo
- Earthworm Jim
- Flashback
- Get Saddam
- Good, Bad, Ugly
- Holiday Lemmings
- Incredible Machine 1, The
- Incredible Machine 2, The
- Jazz Jackrabbit Holiday Hare
- Jump 'n Bump
- Jurassic Park
- Lost Vikings, The
- Magic Pockets
- Mario and Luigi
- Master Blaster
- Moraffs Monster Memory
- Moraffs Super Entrap
- Pinball Fantasies
- Pinball Illusions
- Prehistorik 2
- Ravenloft - Stone Prophet
- Ravenloft - Strahd's Possession
- Scorched Earth
- Summoning, The
- System Shock
- Threat
- Three Kingdoms
- Xiphos
- Xmas Lemmings 1991
- Xmas Lemmings 1992
- Yo! Joe!

## Test sheet

Including details about the upscaled image captures.

- PDF version: [DOSBox Staging image capture upscaling.pdf](https://github.com/dosbox-staging/dosbox-staging/files/12060433/DOSBox.Staging.image.capture.upscaling.pdf)
- ODS spreadsheet:  [DOSBox Staging image capture upscaling.ods.zip](https://github.com/dosbox-staging/dosbox-staging/files/12060462/DOSBox.Staging.image.capture.upscaling.ods.zip)

image

**Comment by johnnovak** [comment]
## Effects of pixel-doubling

Open both in two browser tabs and switch between them at 100% magnification to appreciate the differences.

### With pixel-doubling

EGA 320x200 rendered with pixel-doubling, so 640x200 is sent to the shader. The resulting image looks like a typical sharp EGA monitor.

image


### With no pixel-doubling

EGA 320x200 rendered with _no_ pixel-doubling, so 3200x200 is sent to the shader. The result looks like a cheap 15kHz monitor or a consumer TV set (too much horizontal blur; EGA monitors were meant for business applications and were capable of displaying sharp 640x200 text and graphics).

image

**Comment by johnnovak** [comment]
Shit's building, yo! 😎🤘🏻 

image

**Comment by johnnovak** [comment]
Was a bit worried that the recent rebases and changes from main might have broken ReelMagic, so I've just quickly tested it with Return to Zork. Yep, it's still fine 😎 

image

**Comment by kcgen** [comment]
This is an outstanding cleanup and improvement, @johnnovak!

I haven't been able to review it, but wanted to mention that I've done prior spot-checks against miscelaneous PCjr, Tandy, CGA, and VESA games -- and all are producing correct aspect and screen sizes.

File: include/vga.h [hunk 1]
```diff
@@ -35,29 +35,82 @@
...
 	M_CGA2 = 1 << 0,
+
+	// 320x200 4-colour CGA mode on EGA & VGA
 	M_CGA4 = 1 << 1,
+
+	// 640x480 monochrome EGA mode
+	// All 16-colour EGA and VGA modes
 	M_EGA = 1 << 2,
+
+	// 320x200 256-colour "chunky" or "chanined" VGA mode (mode 13h)
```
Comment (kklobe) [inline include/vga.h:57]: "chained"?

File: include/vga.h [hunk 2]
```diff
@@ -714,7 +767,7 @@ struct VGA_Type {
 	VGA_Memory mem = {};
 	// this is assumed to be power of 2
 	uint32_t vmemwrap = 0;
-	 // memory for fast (usually 16-color) rendering, 
+	 // memory for fast (usually 16-color) rendering,
```
Comment (kklobe) [inline include/vga.h:909]: colour

Comment (kcgen) [reply include/vga.h:909]: kklobe suggesting British/International spelling? Team-mode, engaged! :rocket: :smiley:

File: src/hardware/vga_draw.cpp [hunk 3]
```diff
@@ -1905,505 +1974,616 @@ void VGA_SetupDrawing(uint32_t /*val*/)
...
+		        result in width-doubling the 640x200 mode. A
+		        simple 'width < 640' check will suffice instead
+		        until someone investigates this...
+
+		        if (machine == MCH_TANDY) {
+		        	double_width = (vga.tandy.mode.is_tand_640_dot_graphics) == 0;
+				} else { double_width =
+					(vga.tandy.mode.is_high_bandwidth) == 0;
+		        }
+		*/
```
Comment (johnnovak) [inline src/hardware/vga_draw.cpp:2222]: Comment for self: outdated comment, in the end I got rid of the `width < 640` check; so this should be removed.

File: src/hardware/vga_draw.cpp [hunk 4]
```diff
@@ -1486,255 +1491,411 @@ static std::pair<Fraction, Fraction> determine_pixel_size(const uint32_t htotal,
...
+		vert.blanking_end++;
+
 		if (svga.get_clock) {
 			clock = svga.get_clock();
 		} else {
 			switch ((vga.misc_output >> 2) & 3) {
-			case 0:	
-				clock = (machine==MCH_EGA) ? 14318180 : 25175000;
+			case 0:
+				clock = (machine == MCH_EGA) ? CgaPixelClockHz : 25175000;
```
Comment (kklobe) [inline src/hardware/vga_draw.cpp:1605]: can the 2517500 be de-magic-numbered as well?

File: src/hardware/vga_draw.cpp [hunk 5]
```diff
@@ -1486,255 +1491,411 @@ static std::pair<Fraction, Fraction> determine_pixel_size(const uint32_t htotal,
...
 				break;
 			case 1:
 			default:
-				clock = (machine==MCH_EGA) ? 16257000 : 28322000;
+				clock = (machine == MCH_EGA) ? 16257000 : 28322000;
```
Comment (kklobe) [inline src/hardware/vga_draw.cpp:1609]: more magic numbers.

File: src/hardware/vga_draw.cpp (hunk 4 above)
Comment (johnnovak) [reply src/hardware/vga_draw.cpp:1605]: Sure, if you find out what it stands for and tell me 😄

File: src/hardware/vga_draw.cpp (hunk 5 above)
Comment (johnnovak) [reply src/hardware/vga_draw.cpp:1609]: Found 16257000:

image

Source (page 165):
https://ia801904.us.archive.org/6/items/System_BIOS_for_IBM_PC_XT_AT_Computers_and_Compatibles/System_BIOS_for_IBM_PC_XT_AT_Computers_and_Compatibles.pdf

Comment (johnnovak) [reply src/hardware/vga_draw.cpp:1609]: `VgaFreq1`... thanks? 😅 
You can help tracking these numbers down, I can only use Google search...

Source: https://9p.io/sources/patch/applied/vgaradeon/vga.h

image

File: src/hardware/vga_draw.cpp (hunk 4 above)
Comment (kklobe) [reply src/hardware/vga_draw.cpp:1605]: You're the VGA expert! :grin:

Looks like the clock for 640x(350/400/480) modes? http://tinyvga.com/vga-timing

Comment (johnnovak) [reply src/hardware/vga_draw.cpp:1605]: VGA 640x400 @ 70Hz pixel clock, seems like it.

Source: http://tinyvga.com/vga-timing/640x400@70Hz

image

File: src/hardware/vga_draw.cpp (hunk 5 above)
Comment (johnnovak) [reply src/hardware/vga_draw.cpp:1609]: Looks like the names VgaDotClock 1 & 2 will just do 🤷🏻 

> For example, VGA has 2 video clocks available, 25.175Mhz and 28.322Mhz. 
Ref: http://mirrors.arcadecontrols.com/VGATV/pwp.netcabo.pt/pscoelho/vgatv/faq2.html

image
Ref: http://javiervalcarce.eu/html/vga-signal-format-timming-specs-en.html

Also (search for "28.322"):
https://www.ardent-tool.com/video/VGA_Video_Modes.html
http://gunkies.org/wiki/VGA

File: src/hardware/vga_draw.cpp (hunk 3 above)
Comment (johnnovak) [reply src/hardware/vga_draw.cpp:2222]: Actually, it has to stay. I'm doing `width < 640` later when I add the pixel-doubling stuff, so the comment is still relevant.

**Comment by kklobe** [comment]
All right - love this PR! I have gone through commit-by-commit, and then I just started at the top of `vga_draw.cpp` and worked my way down, slowly. I'm able to much more easily follow what's going on with many of your cleanups and refactorings. I don't really have any substantial changes to request or overall architectural feedback at this point. I gave it a run through with several Sierra games using `fakelottes` and, wow, even my old eyes can see the difference for the 320x200 modes.

Bravo!

**Comment by johnnovak** [comment]
> All right - love this PR! I have gone through commit-by-commit, and then I just started at the top of `vga_draw.cpp` and worked my way down, slowly. I'm able to much more easily follow what's going on with many of your cleanups and refactorings. I don't really have any substantial changes to request or overall architectural feedback at this point. I gave it a run through with several Sierra games using `fakelottes` and, wow, even my old eyes can see the difference for the 320x200 modes.
> 
> Bravo!

Thanks for the review man, glad you're enjoying it! 😄 Yeah CGA/EGA games look a lot closer now to a real monitor; previously the lack of pixel-doubling made the CRT-shaded output look like a really cheap 15kHz monitor or a consumer TV... which can be cool, but EGA monitors were tack sharp.

**Comment by johnnovak** [comment]
I de-magic'ed the pixel clock constants as suggested @kklobe 
https://github.com/dosbox-staging/dosbox-staging/pull/2654/commits/d985e60a6dbed627fb53c1a2f260716ae96f1068

If you're happy @kklobe & @kcgen , let's get this merged! 🚀

**Comment by johnnovak** [comment]
> Another massive improvement, @johnnovak!

Thanks 😊 I'll clean up the PVS Studio warnings by merging https://github.com/dosbox-staging/dosbox-staging/pull/2715 first, then I'll rebase, and hopefully it will be all green again then & ready for merging.

**Comment by kcgen** [comment]
Three new PVS studio bugs are fixed in this PR, too :+1: !
`Total: 267 issues (out of 270 allowed)`

Can ratchet the max down in `.github/workflows/pvs-studio.yml` (near the bottom, `MAX_BUGS:`).

**Comment by johnnovak** [comment]
OFF MY SHOULDER AT LAST! 😎 

https://media.giphy.com/media/aNTmc4keX4Fva/giphy.gif

**Comment by kcgen** [comment]
:shipit:

**Pull request merged**